eclair-core/src/main/scala/fr/acinq/eclair/payment/send/Recipient.scala

```diff
@@ -166,15 +166,18 @@ case class BlindedRecipient(nodeId: PublicKey,
 }
 
 object BlindedRecipient {
-  def apply(invoice: Bolt12Invoice, totalAmount: MilliSatoshi, expiry: CltvExpiry, customTlvs: Set[GenericTlv]): BlindedRecipient = {
+  def apply(invoice: Bolt12Invoice, totalAmount: MilliSatoshi, expiry: CltvExpiry, customTlvs: Set[GenericTlv]): Option[BlindedRecipient] = {
```

Comment:
I think we can do better: it's a bit weird here to return an `Option` to tell the caller that they should have resolved the compact paths first. We can instead simply ask them to provide resolved paths in the arguments, which is more explicit:

```scala
  /**
   * @param paths the caller must resolve the scid of compact blinded paths, otherwise they will be ignored.
   */
  def apply(invoice: Bolt12Invoice, paths: Seq[(Sphinx.RouteBlinding.BlindedRoute, OfferTypes.PaymentInfo)], totalAmount: MilliSatoshi, expiry: CltvExpiry, customTlvs: Set[GenericTlv]): BlindedRecipient = {
    val blindedHops = paths.map {
      case (path, paymentInfo) =>
        // We don't know the scids of channels inside the blinded route, but it's useful to have an ID to refer to a
        // given edge in the graph, so we create a dummy one for the duration of the payment attempt.
        val dummyId = ShortChannelId.generateLocalAlias()
        BlindedHop(dummyId, path, paymentInfo)
    }
    BlindedRecipient(invoice.nodeId, invoice.features, totalAmount, expiry, blindedHops, customTlvs)
  }
```

If we do that, we can also revert all the changes made to the `PaymentInitiator`: the resolution of `scid -> nodeId` should instead be done in `OfferPayment`, once we've received the invoice, and without modifying it! The only change that will be needed in `PaymentInitiator` is to update `SendPaymentToNode` to contain a new field `blindedPaths: Seq[(BlindedRoute, PaymentInfo)]`. I really wasn't a fan of the added complexity and confusion in `PaymentInitiator`, I think this will be much cleaner that way?

Reply:
If we do it in `OfferPayment`, it means we won't be able to pay a Bolt12 invoice provided by the user which is a nice feature.

Reply:
Can you explain why it would be a nice feature? I thought the goal with Bolt 12 was on the contrary to hide the invoice from the user, and directly pay an offer instead.

Even if we were to allow directly paying a Bolt 12 invoice, the resolution of the blinded paths could still be done in the API handler before calling the `PaymentInitiator`, since the flow will likely be that you spin up a one-time actor, give it the invoice, and have it return the list of resolved blinded paths. The current state of the `PaymentInitiator` isn't satisfactory, it has become quite messy with the changes in the latest commit...

Reply:
I've moved it to `OfferPayment`.